configure test for db_env_create fails

## Expected Behaviour

I'm trying to build neomutt 20220429 in pkgsrc on NetBSD with bdb 4 (actually trying to fix the bdb option for the package).
db.h is picked up, but the test for db_env_create fails:
```
ld: /tmp//ccDkYIYE.o: in function `main':
conftest__.c:(.text.startup+0x5): undefined reference to `db_env_create'
child process exited abnormally
============
The failed code was:
extern void db_env_create(void);
int main(void) {
db_env_create();
return 0;
}
============
Failed: gcc -O2 -fPIC -D_FORTIFY_SOURCE=2 -I/usr/include/krb5 -I/usr/include -I/usr/pkg/include -D_XOPEN_SOURCE_EXTENDED=1 -I/usr/pkg/include/ncurses -I/usr/pkg/include/db4 -I/usr/pkg/include/glib-2.0 -I/usr/pkg/include/gio-unix-2.0 -I/usr/pkg/lib/glib-2.0/include -I/usr/pkg/include/lua-5.4 -I/usr/include/readline -std=c99 -fno-delete-null-pointer-checks -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -DNCURSES_WIDECHAR -Wl,-zrelro -L/usr/lib -Wl,-R/usr/lib -L/usr/pkg/lib -Wl,-R/usr/pkg/lib -std=c99 -O2 -fPIC -D_FORTIFY_SOURCE=2 -I/usr/include/krb5 -I/usr/include -I/usr/pkg/include -D_XOPEN_SOURCE_EXTENDED=1 -I/usr/pkg/include/ncurses -I/usr/pkg/include/db4 -I/usr/pkg/include/glib-2.0 -I/usr/pkg/include/gio-unix-2.0 -I/usr/pkg/lib/glib-2.0/include -I/usr/pkg/include/lua-5.4 -I/usr/include/readline -std=c99 -fno-delete-null-pointer-checks -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/usr/pkg/include/db4 -I/include conftest__.c -o conftest__ -ldb-4.8 -L/lib -L/usr/pkg/lib/db4 -lidn -lssl -lcrypto -lncursesw -lnotmuch -llua5.4 -lsasl2 -lgpgme -L/usr/pkg/lib -lgpgme -Wl,-R/usr/pkg/lib -lassuan -lgpg-error -L/usr/pkg/lib -lgpg-error -lintl -liconv
ld: cannot find -ldb-4.8
child process exited abnormally
============
The failed code was:
extern void db_env_create(void);
int main(void) {
db_env_create();
return 0;
}
============
```

which results in a build with only bdb failing at configure time:
```
Checking for BerkeleyDB in /usr/pkg/include/...no
Checking for BerkeleyDB in /usr/pkg/include/db-5-3...no
Checking for BerkeleyDB in /usr/pkg/include/db53...no
Checking for BerkeleyDB in /usr/pkg/include/db5.3...no
Checking for BerkeleyDB in /usr/pkg/include/db-5...no
Checking for BerkeleyDB in /usr/pkg/include/db5...no
Checking for BerkeleyDB in /usr/pkg/include/...no
Checking for BerkeleyDB in /usr/pkg/include/db-6-2...no
Checking for BerkeleyDB in /usr/pkg/include/db62...no
Checking for BerkeleyDB in /usr/pkg/include/db6.2...no
Checking for BerkeleyDB in /usr/pkg/include/db-6...no
Checking for BerkeleyDB in /usr/pkg/include/db6...no
Checking for BerkeleyDB in /usr/pkg/include/...no
Checking for BerkeleyDB in /usr/pkg/include/db-4-8...no
Checking for BerkeleyDB in /usr/pkg/include/db48...no
Checking for BerkeleyDB in /usr/pkg/include/db4.8...no
Checking for BerkeleyDB in /usr/pkg/include/db-4...no
Checking for BerkeleyDB in /usr/pkg/include/db4...yes
Checking for db.h...ok
Checking libs for db_env_create...no
no
Error: Unable to find BerkeleyDB
Try: 'configure --help' for options
*** Error code 1

Stop.
```

## How often does this happen?

- Always

I can provide more details. So far it seems to me as if the bdb piece of our package's Makefile has some old autoconfigure parts left over, the part about BDB_ :
```
### Header cache
###
.if !empty(PKG_OPTIONS:Mtokyocabinet)
.include "../../databases/tokyocabinet/buildlink3.mk"
CONFIGURE_ARGS+=        --tokyocabinet
CONFIGURE_ARGS+=        --disable-gdbm
CONFIGURE_ARGS+=        --disable-bdb
.else
BDB_ACCEPTED=           db4 db5
BUILDLINK_TRANSFORM+=   l:db:${BDB_TYPE}
.include "../../mk/bdb.buildlink3.mk"
CONFIGURE_ARGS+=        --bdb
CONFIGURE_ARGS+=        --disable-gdbm
# BDB_INCLUDE_DIR_ and BDB_LIB_DIR don't have to be particularly accurate
# since the real -I and -L flags are added by buildlink already.
CONFIGURE_ENV+=         BDB_INCLUDE_DIR=${BDBBASE}/include
CONFIGURE_ENV+=         BDB_LIB_DIR=${BDBBASE}/lib
CONFIGURE_ENV+=         BDB_LIB=${BDB_LIBS:S/^-l//:M*:Q}
.endif
```

Your `config.log` shows:

> ld: cannot find -ldb-4.8

If `configure` can't find them, adding `--with-bdb=path` might help (Location of BerkeleyDB)

--

In `auto.def` we do our best to find what the libraries are called and where they live:

https://github.com/neomutt/neomutt/blob/9120e744af1e7289680411072f9dcc9e52e9da9e/auto.def#L942-L953

Checking for db.h...ok                                                                                                                                                                         
Checking libs for db_env_create...-ldb4-4.8 

this worked, thanks!

(now the build fails with:
gmake: *** No rule to make target 'all-data', needed by 'all'.  Stop.
gmake: *** Waiting for unfinished jobs....

but I am pretty sure this is due to my incomplete cherry-picking for the configure phase from master to latest release.

> (now the build fails with: gmake: *** No rule to make target 'all-data', needed by 'all'. Stop. gmake: *** Waiting for unfinished jobs....
> 
> but this is due to my incomplete cherry-picking for the configure phase from master to latest release.

yes, this is most likely due to #3433

--- a/auto.def
+++ b/auto.def
@@ -941,28 +941,16 @@ if {[get-define want-pcre2]} {
 ###############################################################################
 # Header cache - bdb
 if {[get-define want-bdb]} {
-  set bdb_versions [opt-val with-bdb-version { 5.3 6.2 4.8 }] ;# Will be checked in order
-  set bdb_prefix [opt-val with-bdb $prefix]
-
-  foreach ver $bdb_versions {
-    lassign [split $ver .] maj min
-    # This is ugly, but it allows us to not have an inner loop
-    lappend bdb_majors   $maj $maj         $maj           $maj        $maj    $maj
-    lappend bdb_minors   $min $min         $min           $min        $min    $min
-    lappend bdb_exploded ""   db-$maj-$min db${maj}${min} db$maj.$min db-$maj db$maj
-  }
-
-  foreach maj $bdb_majors min $bdb_minors ver $bdb_exploded {
-    set ver_inc_dir $bdb_prefix/include/$ver
-    set ver_lib_dir $bdb_prefix/$::libdir_tail/$ver
+  proc find_bdb {bdb_prefix maj min path_tail lib_names} {
+    set ver_inc_dir $bdb_prefix/include/$path_tail
+    set ver_lib_dir $bdb_prefix/$::libdir_tail/$path_tail
     set ver_inc_file $ver_inc_dir/db.h
-    set ver_lib_file  db-$maj.$min
 
     # File exists?
     msg-checking "Checking for BerkeleyDB in $ver_inc_dir..."
     if {![file exists $ver_inc_file]} {
       msg-result "no"
-      continue
+      return 0
     }
 
     # Version is coherent?
@@ -971,23 +959,41 @@ if {[get-define want-bdb]} {
     if {$inc_maj eq {}   || $inc_min eq {} ||
         $inc_maj != $maj || $inc_min != $min} {
       msg-result "no (expecting $maj.$min, got $inc_maj.$inc_min)"
-      continue
+      return 0
     }
     msg-result "yes"
 
     # Can link?
-    cc-with [list -libs -L$ver_lib_dir -cflags -I$ver_inc_dir] {
-      if {![check-inc-and-lib bdb {} db.h db_env_create db-$maj.$min]} {
-        msg-result "no"
-        continue
+    foreach lib_name $lib_names {
+      cc-with [list -libs -L$ver_lib_dir -cflags -I$ver_inc_dir] {
+        check-inc-and-lib bdb {} db.h db_env_create $lib_name
+      }
+      if {[have-feature bdb]} {
+        define-append CFLAGS  -I$ver_inc_dir
+        define-append LDFLAGS -L$ver_lib_dir
+        define-append LIBS    -l$lib_name
+        define-append HCACHE_BACKENDS "bdb"
+        define USE_HCACHE
+        return 1
       }
     }
-    define-append CFLAGS  -I$ver_inc_dir
-    define-append LDFLAGS -L$ver_lib_dir
-    define-append LIBS    -ldb-$maj.$min
-    define-append HCACHE_BACKENDS "bdb"
-    define USE_HCACHE
-    break
+    return 0
+  }
+
+  set bdb_versions [opt-val with-bdb-version { 5.3 6.2 4.8 }] ;# Will be checked in order
+  set bdb_prefix [opt-val with-bdb $prefix]
+
+  foreach ver $bdb_versions {
+    lassign [split $ver .] maj min
+    foreach path_tail [list "" db-$maj-$min db${maj}${min} db$maj.$min db-$maj db$maj] {
+      set lib_names [list db$maj-$maj.$min db$maj-$maj db-$maj.$min db-$maj]
+      if {[find_bdb $bdb_prefix $maj $min $path_tail $lib_names]} {
+        break
+      }
+    }
+    if {[have-feature bdb]} {
+      break
+    }
   }
 
   if {![have-feature bdb]} {

